Improved integrity errors handling

warrkan opened the pull request:
<!-- Contributing guide: https://github.com/open-edge-platform/training_extensions/blob/develop/CONTRIBUTING.md -->

### Summary

Resolves #4777

<!--
Please add a summary of changes. You may use Copilot to auto-generate the PR description but please consider
including any other relevant facts which Copilot may be unaware of (such as design choices and testing procedure).

Add references to the relevant issues and pull requests if any like so:

Resolves #111 and #222.
Depends on #1000 (for series of dependent commits).
-->

### How to test

<!-- Describe the testing procedure for reviewers, if changes are
not fully covered by unit tests or manual testing can be complicated. -->

### Checklist

<!-- Put an 'x' in all the boxes that apply -->

- [ ] The PR title and description are clear and descriptive
- [x] I have manually tested the changes
- [x] All changes are covered by automated tests
- [x] All related issues are linked to this PR (if applicable)
- [ ] Documentation has been updated (if applicable)

warrkan committed:
Improved integrity errors handling

leoll2 commented on application/backend/app/services/project_service.py:
```diff
@@ -29,7 +31,10 @@ def __init__(self, data_dir: Path) -> None:
+        try:
+            return self._persistence.create(project)
+        except UniqueConstraintIntegrityError:
+            raise DuplicateLabelsError
```
The project creation endpoint allows the client to provide an id. It's a bit of an edge case, but it may happen that the client-provided id conflicts with an existing project and a `PrimaryKeyIntegrityError` is raised here. In such case, does the API return 409 with a clear error message? You can test it by calling `POST /projects` twice with the same id in the payload.

warrkan replied:
Yes, if you try to create a project with the existing ID, you'll get a proper error:
<img width="557" height="306" alt="image" src="https://github.com/user-attachments/assets/2863dfd6-38ea-47c3-a70d-7aa2423a9e05" />

leoll2 reviewed (dismissed):
LGTM, one minor comment

warrkan committed:
Merge branch 'develop' into romanove/4777_exceptions_handling

warrkan committed:
Fixed style check

itallix commented on application/backend/app/api/endpoints/sinks.py:
```diff
@@ -97,7 +102,7 @@ async def create_sink(
     try:
         return configuration_service.create_sink(sink_config)
-    except ResourceAlreadyExistsError as e:
+    except ResourceWithNameAlreadyExistsError as e:
```
We should handle the case where ids are provided as well, since all resources allow passing an ID during creation.

warrkan replied:
It will be handled by https://github.com/open-edge-platform/training_extensions/pull/4812/files#diff-831489a76e6762f2886866f1bdbffe2953a1fc9dbd49ee5085ab59c2176d74f5R122

itallix replied:
```suggestion
    except (ResourceWithNameAlreadyExistsError, ResourceWithIdAlreadyExistsError) as e
```

warrkan committed:
Split project & label management on the business logic level

warrkan committed:
Merge branch 'develop' into romanove/4777_exceptions_handling

leoll2 commented on application/backend/app/services/data_collect/data_collector.py:
```diff
@@ -107,7 +107,11 @@ def collect(
         )
         with get_db_session() as session:
-            dataset_service = DatasetService(self.data_dir, session)
+            label_service = LabelService(db_session=session)
+            project_service = ProjectService(data_dir=self.data_dir, db_session=session, label_service=label_service)
+            dataset_service = DatasetService(
+                data_dir=self.data_dir, db_session=session, project_service=project_service
+            )
```
Would it be possible to simplify this code by letting the `DatasetService` instantiate `ProjectService` if not provided by argument? Similarly, `ProjectService` could instantiate a `LabelService` if not passed explicitly.

warrkan replied:
I wouldn't do it, let's keep instances instantiation a responsibility of DI component. Otherwise it can become a mess

warrkan replied:
@itallix what do you think?

itallix replied:
The current approach with DI is the correct pattern and should be maintained. Services should focus on their business logic, not instantiating their dependencies. 
We can extract factories when a number of dependencies is required like here.

BTW, we still violate this pattern when it comes to repositories - possible task for a future refactor.

warrkan committed:
Fixed DB seeding

leoll2 commented on application/backend/app/services/__init__.py:
```diff
@@ -1,16 +1,24 @@
 # SPDX-License-Identifier: Apache-2.0
-from .active_model_service import ActiveModelService
+
+from .base import (  # isort: skip
```
Sorting the imports manually through `# isort: skip` allows to avoid circular imports, but it becomes harder to maintain as the number of imports grows. A possible alternative, imho better, is to use _relative_ imports within the module (`app.services` in this case), as in the following example:

<img width="726" height="272" alt="image" src="https://github.com/user-attachments/assets/650c5c34-08ae-491f-9f08-4af681bf2f51" />

If you agree with this approach, I can push a commit to fix all the imports in the services module.

leoll2 replied:
Pushed fix -> [02a43b4](https://github.com/open-edge-platform/training_extensions/pull/4812/commits/02a43b409145bb7e663cac09bc32c4a258e609d5)

warrkan replied:
Thanks @leoll2 !

leoll2 committed:
use relative imports within the service module

this is an alternative approach to avoid circular imports, easier to maintain because it doesn't require to manually sort the imports in the __init__ file

itallix commented on application/backend/app/services/active_pipeline_service.py:
```diff
@@ -75,7 +78,7 @@ def _load_app_config(self) -> None:
             project = pipeline.project
             if project is not None:
-                self._project = ProjectMapper.to_schema(project)
+                self._project = ProjectMapper.to_schema(project, LabelService(db).list_all(UUID(project.id)))
```
I believe we've reached a point where view objects no longer align well with our internal service needs. While these objects work fine when used by routers, they're not ideal for internal operations - particularly in stateful services like `ActiveModelService` and `ActivePipelineService` (e.g., labels are meaningless within the scope of `ActivePipelineService`).

For services that maintain runtime state, we should consider designing proper domain objects tailored to their specific operational requirements. However, this is probably something for future tasks rather than the current PR.

warrkan replied:
I completely agree, and afaik I've raised this topic at some point. In general it's a good idea to separate presentation layer objects and business layer objects.

itallix commented on application/backend/app/api/endpoints/sources.py:
```diff
@@ -108,7 +114,7 @@ def create_source(
     try:
         return configuration_service.create_source(source_create)
-    except ResourceAlreadyExistsError as e:
+    except ResourceWithNameAlreadyExistsError as e:
```
```suggestion
    except (ResourceWithNameAlreadyExistsError, ResourceWithIdAlreadyExistsError) as e:
```

warrkan committed:
Added exceptions handling

itallix commented on application/backend/app/services/dispatchers/filesystem.py:
```diff
@@ -10,7 +10,8 @@
 from app.schemas.sink import FolderSinkConfig, OutputFormat
-from app.services.dispatchers.base import BaseDispatcher
+
+from .base import BaseDispatcher
```
Is there a reason to change this to relative import.

Currently we have three groups of imports:
1. sdk
2. third party libs
3. app deps

This is introducing unnecessary group within app deps (abosolute + relative). I find it convenient for `__init__.py` to expose inners, but probably unnecessary here?

leoll2 replied:
Related comment: https://github.com/open-edge-platform/training_extensions/pull/4812#discussion_r2418911821

When Python modules import classes from sibling modules and these classes are exposed with `__init__.py`, it's very common to run into circular imports. To avoid that, here we use relative imports. It's important to note that relative imports are used only to import other _services_ (i.e. classes defined within this module `app.services`), while for everything else it is still better to use absolute imports.

warrkan commented on application/backend/app/services/mappers/label_mapper.py:
```diff
@@ -15,7 +16,9 @@ def to_schema(label_db: LabelDB) -> Label:
         """Convert Label schema to db model."""
 
-        return LabelDB(**label.model_dump(mode="json"))
+        label_db = LabelDB(**label.model_dump(mode="json"))
```
Label doesn't have `project_id` field, so it cannot be initialized via constructor

leoll2 commented on application/backend/app/services/dataset_service.py:
```diff
@@ -162,36 +164,40 @@ def list_dataset_items(
         """Delete a dataset item by its ID"""
-        repo = DatasetItemRepository(project_id=str(project_id), db=self._db_session)
+        project = self._project_service.get_project_by_id(project_id)
+        repo = DatasetItemRepository(project_id=str(project.id), db=self._db_session)
```
Why do you need to load the `project` here? What's wrong with the old implementation that used `project_id` directly?

warrkan replied:
Because I think it's more correct to indicate that project doesn't exist rather than dataset item is missing

warrkan committed:
Merge branch 'develop' into romanove/4777_exceptions_handling

warrkan commented on application/backend/app/services/configuration_service.py:
```diff
@@ -26,16 +34,46 @@ class PipelineField(StrEnum):
+        except UniqueConstraintIntegrityError:
+            raise ResourceWithNameAlreadyExistsError(ResourceType.SOURCE, item.name)
+
+    def update(self, item: Source, partial_config: dict) -> Source:
+        try:
+            return super().update(item, partial_config)
+        except UniqueConstraintIntegrityError:
+            raise ResourceWithNameAlreadyExistsError(ResourceType.SOURCE, partial_config["name"])
```
If the error is raised, it means that there was an attempt to change source name to existing one. It means that `name` in this case must be present in `partial_config`

warrkan commented on application/backend/app/services/configuration_service.py:
```diff
@@ -26,16 +34,46 @@ class PipelineField(StrEnum):
+        except UniqueConstraintIntegrityError:
+            raise ResourceWithNameAlreadyExistsError(ResourceType.SINK, item.name)
+
+    def update(self, item: Sink, partial_config: dict) -> Sink:
+        try:
+            return super().update(item, partial_config)
+        except UniqueConstraintIntegrityError:
+            raise ResourceWithNameAlreadyExistsError(ResourceType.SINK, partial_config["name"])
```
The same here
